File: tool/uapicodegen/uapicodegen/uapi/internal/schema/ParseErrorType.py
from typing import TYPE_CHECKING, cast
from ...internal.schema.SchemaParseFailure import SchemaParseFailure
from ...internal.types.UError import UError

if TYPE_CHECKING:
    from ...internal.schema.ParseContext import ParseContext


def parse_error_type(path: list[object], error_definition_as_parsed_json: dict[str, object],
                     schema_key: str,
                     ctx: 'ParseContext') -> 'UError':
    from ...UApiSchemaParseError import UApiSchemaParseError
    from ...internal.schema.ParseUnionType import parse_union_type

    parse_failures = []

    other_keys = set(error_definition_as_parsed_json.keys())
    other_keys.discard(schema_key)
    other_keys.discard("///")

    if other_keys:
        for k in other_keys:
            loop_path = path + [k]
            parse_failures.append(SchemaParseFailure(
                ctx.document_name, cast(list[object], loop_path), "ObjectKeyDisallowed", {}))

    if parse_failures:
        raise UApiSchemaParseError(
            parse_failures, ctx.uapi_schema_document_names_to_json)

    error = parse_union_type(path, error_definition_as_parsed_json, schema_key, [], [],
                             ctx)

    return UError(schema_key, error)

File: lib/java/src/main/java/io/github/telepact/internal/schema/ParseErrorType.java
package io.github.telepact.internal.schema;

import static io.github.telepact.internal.schema.ParseUnionType.parseUnionType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.github.telepact.TelepactSchemaParseError;
import io.github.telepact.internal.types.VError;
import io.github.telepact.internal.types.VUnion;

public class ParseErrorType {
    public static VError parseErrorType(
            List<Object> path,
            Map<String, Object> errorDefinitionAsParsedJson,
            String schemaKey,
            ParseContext ctx) {

        List<SchemaParseFailure> parseFailures = new ArrayList<>();

        Set<String> otherKeys = new HashSet<>(errorDefinitionAsParsedJson.keySet());
        otherKeys.remove(schemaKey);
        otherKeys.remove("///");

        if (!otherKeys.isEmpty()) {
            for (String k : otherKeys) {
                List<Object> loopPath = new ArrayList<>(path);
                loopPath.add(k);
                parseFailures
                        .add(new SchemaParseFailure(ctx.documentName, loopPath, "ObjectKeyDisallowed",
                                new HashMap<>()));
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new TelepactSchemaParseError(parseFailures, ctx.telepactSchemaDocumentNamesToJson);
        }

        // Assuming parseUnionType is adapted to Java and returns UError or its

        VUnion error = parseUnionType(path, errorDefinitionAsParsedJson, schemaKey,
                new ArrayList<>(),
                new ArrayList<>(), ctx);

        return new VError(schemaKey, error);
    }
}
